main.go

```diff
@@ -80,212 +76,235 @@ func main() {
-					}
+				isRunning = false
+			case registry.CHANGED:
+				if !isRunning {
+					glog.Infof("Ignored CHANGED update of service with tag %q since it's not running", tagName)
+					continue
 				}
-
-				// do we have new instances to add to the NEG?
-				if len(toAdd) > 0 {
-					if err := client.AttachEndpointsToGroup(toAdd, makeName("neg", serviceGroupName)); err != nil {
-						glog.Errorf("Failed adding endpoints to %s network endpoint group: %v", makeName("neg", serviceGroupName), err)
-					}
+				if err := handleServiceChangedUpdate(client, tagConfig, lock, instances, update); err != nil {
+					glog.Errorf(errorTemplate, update.UpdateType, tagName, err)
 				}
-
-				lock.Unlock()
-			default:
-				continue
 			}
 		case <-done:
-			glog.Warningf("Received termination signal for service with tag %s", tag)
+			glog.Warningf("Received termination signal for service with tag %q", tagName)
 			wg.Done()
 			return
 		}
 	}
 }
 
-// NOTE: Maximum length of resource name is 63 and name must start and end with letter or digit
-func makeName(prefix string, name string) string {
-	n := strings.Join([]string{prefix, name}, "-")
-	if len(n) > maximumNameLength {
-		n = strings.TrimLeft(n[:maximumNameLength], "-")
+func handleServiceNewUpdate(
+	urlMapName string,
+	client cloud.Cloud,
+	tagConfig *tagConfiguration,
+	lock *sync.RWMutex,
+) error {
+	lock.Lock()
+	defer lock.Unlock()
+
+	tagName := tagConfig.Name
+
+	if tagConfig.Mode == REUSE_MODE {
+		glog.Infof("Service with tag %q will use externally managed resources", tagName)
+	} else {
+		negName := tagConfig.NetworkEndpointGroup.Name
+		glog.Infof("Creating network endpoint group %q...", negName)
+		if err := client.CreateNetworkEndpointGroup(negName); err != nil {
+			return fmt.Errorf("Can't create network endpoint group %q: %+v", negName, err)
+		}
+		glog.Infof("Created network endpoint group %q", negName)
+
+		hcName := tagConfig.HealthCheck.Name
+		glog.Infof("Creating health check %q...", hcName)
+		if err := client.CreateHealthCheck(hcName, tagConfig.HealthCheck.Path); err != nil {
+			return fmt.Errorf("Can't create health check %q: %+v", hcName, err)
+		}
+		glog.Infof("Created health check %q", hcName)
+
+		bsName := tagConfig.BackendService.Name
+		glog.Infof("Creating backend service %q...", bsName)
+		if err := client.CreateBackendService(bsName, negName, hcName, tagConfig.BackendService.Affinity, tagConfig.BackendService.CDN); err != nil {
+			return fmt.Errorf("Can't create backend service %q: %+v", bsName, err)
+		}
+		glog.Infof("Created backend service %q", bsName)
+
+		glog.Infof("Updating URL map %q...", urlMapName)
+		if err := client.UpdateURLMap(urlMapName, bsName, tagConfig.BackendService.Host, tagConfig.BackendService.Path); err != nil {
+			return fmt.Errorf("Can't update URL map %q: %+v", urlMapName, err)
+		}
+		glog.Infof("Updated URL map %q", urlMapName)
```

Comment:
I looked through these APIs and they never have retry logic in them, which might make this part to be "flacky" and noise in future monitoring systems. Not sure if it is something we should address though.

Reply:
Retry logic wouldn't be excess here indeed. Currently, we plan to use `REUSE` mode for tags and those resources are created externally but attach/detach endpoints into/from NEGs logic can be extended with retry thing. We can discuss it and solve later out of this PR scope.

Reply:
https://github.com/dffrntmedia/consul-lb-gce/issues/9